components/plurals/src/provider.rs

```diff
@@ -319,3 +319,164 @@ mod ranges {
         pub ranges: ZeroMap<'data, UnvalidatedPluralRange, RawPluralCategory>,
     }
 }
+
+#[cfg(any(doc, feature = "datagen"))]
+use crate::PluralElements;
+use crate::{PluralCategory, PluralOperands, PluralRules};
+use alloc::borrow::Cow;
+use zerovec::ule::AsULE;
+use zerovec::VarZeroVec;
+
+#[derive(Debug, Clone, Copy)]
+#[zerovec::make_ule(PluralCategoryV1ULE)]
+#[repr(u8)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+/// Keys for the different fields of [`PluralElements`].
+pub enum PluralElementsKeysV1 {
+    /// Key for [`PluralElements::zero`]
+    Zero = 0,
+    /// Key for [`PluralElements::one`]
+    One = 1,
+    /// Key for [`PluralElements::two`]
+    Two = 2,
+    /// Key for [`PluralElements::few`]
+    Few = 3,
+    /// Key for [`PluralElements::many`]
+    Many = 4,
+    /// Key for [`PluralElements::explicit_zero`]
+    ExplicitZero = 5,
+    /// Key for [`PluralElements::explicit_one`]
+    ExplicitOne = 6,
+}
+
+#[derive(Debug)]
+#[zerovec::make_varule(PluralElementsFieldV1ULE)]
+#[zerovec::skip_derive(Ord)]
+#[zerovec::derive(Debug)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize),
+    zerovec::derive(Deserialize)
+)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize),
+    zerovec::derive(Serialize)
+)]
+/// A tuple of [`PluralElementsKeysV1`] and [`str`].
+// TODO: Make the str generic
+pub struct PluralElementsFieldV1<'data>(pub PluralElementsKeysV1, pub Cow<'data, str>);
+
+// TODO: Make the str generic
+#[derive(Debug, Clone, PartialEq, yoke::Yokeable, zerofrom::ZeroFrom)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize, databake::Bake))]
+#[cfg_attr(feature = "datagen", databake(path = icu_plurals::provider))]
+#[yoke(prove_covariance_manually)]
+/// A data-struct-optimized representation of [`PluralElements`].
+pub struct PluralElementsV1<'data> {
+    /// Optional entries for categories other than PluralCategory::Other
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub specials: VarZeroVec<'data, PluralElementsFieldV1ULE>,
+    /// The entry for PluralCategory::Other
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub other: Cow<'data, str>,
```

Comment:
Question: This does not implement VarULE, but I think we may want to put this in a VarZeroVec. How do you envision making it VarULE? (Doesn't need to be now, but I'd like to have a path)

Reply:
I thought you didn't want nested `VarZeroVec`s?